Setting up an environment for quantum computing

cryos opened the pull request:
This is a baseline set of Python modules and a QM code (Psi4) to do some
basic quantum computing simulation.

cryos committed:
Setting up an environment for quantum computing

This is a baseline set of Python modules and a QM code (Psi4) to do some
basic quantum computing simulation.

Signed-off-by: Marcus D. Hanwell <marcus.hanwell@kitware.com>

cryos commented:
This was pushed to openchemistry/quantum:dev too @muammar and @wadejong

muammar commented on docker/quantum/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+RUN apt-get update && \
+  apt-get install -y libblas-dev liblapack-dev && \
+  cd /root && \
+  pip install -r requirements.txt && \
```
Before installing the modules in `requirements.txt`, one needs to add the line: 

```
pip install pybind11 && \
```

This avoids a problem when building `projectq` (see line 3 of `requirements.txt`). After doing that change, my image was built and I could test qiskit. 

<img width="1161" alt="Screen Shot 2020-05-21 at 16 58 40" src="https://user-images.githubusercontent.com/954209/82617381-c4534000-9b84-11ea-8fa9-805a6bce0ed9.png">

muammar commented on docker/quantum/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+  mkdir build && \
+  cd build && \
+  cmake ../psi4-1.3.2 && make -j10 install && \
+  rm -rf /usr/local/psi4/share/psi4/samples && \
```
I tried the psi4 inside the image and it's failing with the following: 

```
  root@2fc7f25c76b0:/tmp# /usr/local/psi4/bin/psi4 input.dat
Traceback (most recent call last):
  File "/usr/local/psi4/bin/psi4", line 287, in <module>
    exec(content)
  File "<string>", line 35, in <module>
  File "/usr/local/psi4/lib/psi4/driver/driver.py", line 556, in energy
    wfn = procedures['energy'][lowername](lowername, molecule=molecule, **kwargs)
  File "/usr/local/psi4/lib/psi4/driver/procrouting/proc.py", line 2002, in run_scf
    scf_wfn = scf_helper(name, post_scf=False, **kwargs)
  File "/usr/local/psi4/lib/psi4/driver/procrouting/proc.py", line 1276, in scf_helper
    base_wfn = core.Wavefunction.build(scf_molecule, core.get_global_option('BASIS'))
  File "/usr/local/psi4/lib/psi4/driver/p4util/python_helpers.py", line 115, in _core_wavefunction_build
    basis = core.BasisSet.build(mol, "ORBITAL", basis)
  File "/usr/local/psi4/lib/psi4/driver/p4util/python_helpers.py", line 80, in _pybuild_basis
    mol.to_dict(), key, resolved_target, fitrole, other, return_dict=True, return_atomlist=return_atomlist)
  File "/usr/local/psi4/lib/psi4/driver/qcdb/molecule.py", line 1461, in to_dict
    compare_molrecs(validated_molrec, molrec, 6, 'to_dict', forgive=forgive, verbose=0)
  File "/usr/local/psi4/lib/psi4/driver/qcdb/psiutil.py", line 255, in compare_molrecs
    compare_dicts(xptd, cptd, tol, label, forgive=forgive, verbose=verbose)
  File "/usr/local/psi4/lib/psi4/driver/qcdb/psiutil.py", line 152, in compare_dicts
    raise ImportError("""Python module deepdiff not found. Solve by installing it: `conda install deepdiff -c conda-forge` or `pip install deepdiff`""")

ImportError: Python module deepdiff not found. Solve by installing it: `conda install deepdiff -c conda-forge` or `pip install deepdiff`

Printing out the relevant lines from the Psithon --> Python processed input file:
    ""","cswater")
    core.IO.set_default_namespace("cswater")
    oswater = cswater.clone()
    oswater.set_multiplicity(2)
    oswater.set_molecular_charge(1)
--> e = energy('b3lyp', molecule=cswater)
    core.set_global_option("REFERENCE", "uhf")
    e = energy('b3lyp', molecule=oswater)
    core.set_global_option("REFERENCE", "rks")
    e = energy('b3lyp5', molecule=cswater)
    core.set_global_option("REFERENCE", "uks")
``` 

The example I was trying to run is: https://github.com/psi4/psi4/blob/master/samples/dft-b3lyp/input.dat

muammar commented on docker/quantum/requirements.txt:
```diff
@@ -0,0 +1,9 @@
+jinja2
+cmake
+pint
+pydantic
```
Add `deepdiff` to the list of requirements.

muammar reviewed (commented):
The `projectq` module is failing to build with the error below: 

 ```    
 pp_opts.append("-I%s" % dir)
    File "/tmp/pip-install-x4_s57sy/projectq/setup.py", line 63, in __str__
      import pybind11
  ModuleNotFoundError: No module named 'pybind11'
  ----------------------------------------
  ERROR: Failed building wheel for projectq
```

cjh1 committed:
Install pybind11, needed by projectq

cjh1 committed:
Add deepdiff to requirements.txt, needed by psi4

cjh1 commented:
I have pushed the changes suggested by @muammar ( thanks :smile: ) `psi4` will now run. @cryos It looks like we are missing python script for the entrypoint, does this exist? Or does it need to be written?

cryos commented:
It needs to be written, I was going to base it on what Bert shared in quantum notebooks.

cjh1 commented:
@cryos It looks like there are a couple of notebooks/calculations which one where we planning on using?

cryos commented:
@cjh1 I hadn't decided, I was just going to pick one and get something working to ensure all the pieces were functional. If it runs I think that needs some input from Bert on the next step as they are just examples that may not form the best driver. It might be good to get together and refresh on what we want to get into the driver/container.

wadejong commented:
Yeah, we probably should do some planning on a workflow.


On Tue, Oct 20, 2020 at 11:39 AM Marcus D. Hanwell <notifications@github.com>
wrote:

> @cjh1 <https://github.com/cjh1> I hadn't decided, I was just going to
> pick one and get something working to ensure all the pieces were
> functional. If it runs I think that needs some input from Bert on the next
> step as they are just examples that may not form the best driver. It might
> be good to get together and refresh on what we want to get into the
> driver/container.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenChemistry/mongochemdeploy/pull/107#issuecomment-713059382>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTL24QBT2DMA2BL47WMPHDSLXKMJANCNFSM4NCT5GHA>
> .
>

bnmajor committed:
Merge branch 'quantum-computing' of https://github.com/OpenChemistry/mongochemdeploy into quantum-computing

bnmajor committed:
Update environment

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Provide input description and add examples

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Add calculation code

This calculation code comes from the notebook provided here:
https://github.com/OpenChemistry/quantumnotebooks/blob/master/OpenChemistry-Quantum.ipynb

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Add input parameter definitions for optimization

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Add optimization function

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Call optimization if params provided in input

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Update output

Include Psi4 energy and properly format coordinates

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor commented:
An explanation of the changes made up until a5edc5f can be found in [this comment](https://github.com/OpenChemistry/mongochemdeploy/issues/108#issuecomment-754916686)

cjh1 commented:
@bnmajor Could you provide a notebook showing how this container can be invoked?

bnmajor commented:
@cjh1 Yes I'll work on one

bnmajor committed:
Update output

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Add instructions to register image

Signed-off-by: Brianna Major <brianna.major@kitware.com>

cjh1 committed:
Merge remote-tracking branch 'origin/master' into quantum-computing

cjh1 commented on docker/quantum/README.md:
```diff
@@ -45,3 +45,18 @@ A json description of the image and some of the options may be obtained via:
+    - `tag`: <strong>1.0</strong>
+    - `digest`: <strong>123</strong> (can be any value)
+    - `clusterId`: (from GET /clusters)
+    - `size`: <strong>2</strong> (can be any value)
```
We should be able to do this via the UI @psavery? Also how do we get it registered automatically in the dev environment?

cjh1 replied:
Looks like we needed to add it to this [list](https://github.com/OpenChemistry/openchemistrypy/blob/master/taskflows/taskflows/container_list_taskflow.py#L32).

psavery replied:
@cjh1 Yeah, that's to register it automatically.

You can also manually register it via the girder endpoint at `localhost:8080/api/v1` under "images". If you do this, you can put whatever you want for the digest and the size, the digest is only used to see if the image has already been registered. The size isn't used for anything but display.

cjh1 replied:
@bnmajor Lets add `openchemistry/qiskit` to the list to be added automatically, then I think we can remove this block.

psavery replied:
We might also want to add it [here](https://github.com/OpenChemistry/mongochemclient/blob/a75f70f569960e1a4c3b0a6b464c927b324c36f9/src/components/image-manager/advanced-pull.js#L21-L24), and [here](https://github.com/OpenChemistry/mongochemclient/blob/a75f70f569960e1a4c3b0a6b464c927b324c36f9/src/utils/calculations.js#L4-L17)

cjh1 commented on docker/quantum/README.md:
````diff
@@ -0,0 +1,62 @@
+docker image:
+
+```
+docker build -t quantum:1.0 .
````
Lets tag the image `openchemistry/qiskit`

cjh1 reviewed (commented):
@bnmajor This works for me with the sample notebook with the inchikey change. Looking good, couple of inline comments.

cjh1 commented:
[Here](https://github.com/OpenChemistry/mongochemserver/blob/fff5b7c698afbd7e3f1489b40057ae81582a0e54/girder/notebooks/notebooks/notebooks/Quantum.ipynb) is the sample notebook.

cjh1 commented:
Any feedback @cryos @wadejong?  We are going to merge soon.

bnmajor committed:
Rename container

Signed-off-by: Brianna Major <brianna.major@kitware.com>

bnmajor committed:
Remove instructions for manually adding image

Signed-off-by: Brianna Major <brianna.major@kitware.com>
